### ISSUE TITLE
Bug: Syntax error handling in console.html

### ISSUE THREAD

**Issue by hoodmane** [pyodide/pyodide#1286]
Syntax errors in console.html are not printed immediately, only after the next command is submitted. E.g., type `1+<ENTER>` into console.html.

I think this was introduced in #1125. I investigated but I can't figure out why it's happening. Any ideas @casatir?

**Comment by casatir** [comment]
Thanks @hoodmane for catching this. I'll investigate.

**Comment by casatir** [comment]
I am pretty sure it was working after #1155.

**Comment by hoodmane** [comment]
I will check (unfortunately, #1155 was before the emscripten update was merged so I have to rebuild everything).

**Comment by casatir** [comment]
Doing

```js
pyconsole.stderr_callback = $.terminal.active().error
```

in JS console makes it work (with newline of course).

**Comment by casatir** [comment]
So it seems related to newline stuff since it works with 

``` js
pyconsole.stderr_callback = $.terminal.active().echo
```

but not with 

``` js
pyconsole.stderr_callback = (s) => $.terminal.active().echo(s, {newline : false})
```

**Comment by hoodmane** [comment]
Does that prevent `sys.except_hook` from happening? I would guess that doing that would cause most errors to be printed twice?

**Comment by casatir** [comment]
This is `code.py` 's `InteractiveConsole` that manage this. `sys.exepthook` seems to not be called in our case since https://github.com/python/cpython/blob/f82578ace103ec977cec3424b20e0b5f19cf720e/Lib/code.py#L123

**Comment by casatir** [comment]
I suspect something like syntax error occurs when terminal is suspended which `echo_newline.js` may dislike.

**Comment by hoodmane** [comment]
Okay it definitely does work after #1155.

**Comment by casatir** [comment]
I think I got it. In `console.html`, we continuously set the prompt:

```js
async function interpreter(command) {
            // multiline should be splitted (usefull when pasting)
            term.pause();
            for( const c of command.split('\n') ) {
                const prompt = pyconsole.push(c) ? ps2 : ps1;
                term.set_prompt(prompt);
                await pyconsole.run_complete;
            }
            term.resume();
        }
```
which disturb `echo_newline.js` since it also set the prompt...

**Comment by hoodmane** [comment]
`echo_newline.js` uses a really awful hack. I might try to make an improvement upstream.

**Comment by casatir** [comment]
I remember I've seen this when trying to make `echo_newline.js` work.

**Comment by casatir** [comment]
Partially reverting #1274 restoring my `term.echo_no_newline` function plus

```js
pyconsole.stdout_callback = term.echo_no_newline;
pyconsole.stderr_callback = (s) => term.echo_no_newline(`[[;red;]${$.terminal.escape_brackets(s)}\u200B]`);
```

makes it work without loading `echo_newline.js`.

**Comment by casatir** [comment]
I can submit a PR with this correction or we can wait for upstream fix. WDYT?

**Comment by hoodmane** [comment]
I opened a PR: https://github.com/jcubic/jquery.terminal/pull/652

**Comment by casatir** [comment]
The why it happens only for syntax errors (and not runtime errors) can be understood with this three lines:

```js
    const prompt = pyconsole.push(c) ? ps2 : ps1;
    term.set_prompt(prompt);
    await pyconsole.run_complete;
```

Execution steps for runtime errors (or even code exec without any error) are:
1. `pyconsole.push(c)` returns `false` (code is not yet executed)
2. prompt is set to `>>>`
3. code is executed (via the `run_complete` promise), error is printed to `stderr`
4. JQuery Terminal shows it by setting the prompt.

While for syntax error, we have:
1. `pyconsole.push(c)` prints syntax error to `stderr` (and return `false`)
2. JQuery Terminal shows it by setting the prompt
3. prompt is erased by our call to `set_prompt`
4. `run_complete` is a dummy promise so the `await` has no effect.

**Comment by hoodmane** [comment]
So as an update, my jquery terminal PR has been reviewed and has received positive feedback so it will probably be merged soon. 
It should make the behavior of `echo` / `error` with `newline : false` completely consistent with the behavior with `newline : true` and fix the various issues we've had with it (any interaction with `set_prompt` has been removed, `term.error` will get the color right, etc).
https://github.com/jcubic/jquery.terminal/pull/654

It looks like releases of `jquery.terminal` are fairly frequent (about one every two weeks recently) and the most recent release was 14 days ago so it probably won't be too long before this change is included into a release.

**Comment by hoodmane** [comment]
Closed by #1480. `jquery.terminal` just released a version with my PR included today!
